From b17cd165979e0fa8240d6c42a57694870e0e249e Mon Sep 17 00:00:00 2001
From: William Douglas <william.douglas@intel.com>
Date: Wed, 8 Nov 2023 13:55:40 -0800
Subject: [PATCH 3/5] common: Support a stateless /etc/lightdm/users.conf

With this change we can support both the admin-defined users.conf, as well
as a vendor-provided users.conf in /usr/share/defaults/lightdm/. This will
permit vendor configurations to provide the default sane settings, in
instances where it is not desirable to provide AccountsService in a lighter
deployment.

In the instance that the admin defined file exists, it will be used instead
of the vendor-config.

Signed-off-by: Ikey Doherty <michael.i.doherty@intel.com>
Signed-off-by: Miguel Bernal Marin <miguel.bernal.marin@linux.intel.com>
Signed-off-by: William Douglas <william.douglas@intel.com>
---
 common/user-list.c | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/common/user-list.c b/common/user-list.c
index fc04a12..53027a5 100644
--- a/common/user-list.c
+++ b/common/user-list.c
@@ -165,6 +165,7 @@ G_DEFINE_TYPE (CommonSession, common_session, G_TYPE_OBJECT)
 
 #define PASSWD_FILE      "/etc/passwd"
 #define USER_CONFIG_FILE "/etc/lightdm/users.conf"
+#define VENDOR_USER_CONFIG_FILE "/usr/share/defaults/lightdm/users.conf"
 
 static CommonUserList *singleton = NULL;
 
@@ -323,13 +324,21 @@ load_passwd_file (CommonUserList *user_list, gboolean emit_add_signal)
 {
     CommonUserListPrivate *priv = common_user_list_get_instance_private (user_list);
 
-    g_debug ("Loading user config from %s", USER_CONFIG_FILE);
+    g_autofree gchar *user_config_file = NULL;
+    /* Load /etc first, otherwise fallback to /usr/share/defaults *
+     * for stateless operation                                    */
+    if (g_file_test(USER_CONFIG_FILE, G_FILE_TEST_EXISTS))
+        user_config_file = g_strdup (USER_CONFIG_FILE);
+    else
+        user_config_file = g_strdup (VENDOR_USER_CONFIG_FILE);
+
+    g_debug ("Loading user config from %s", user_config_file);
 
     g_autoptr(GKeyFile) config = g_key_file_new ();
     g_autoptr(GError) error = NULL;
-    g_key_file_load_from_file (config, USER_CONFIG_FILE, G_KEY_FILE_NONE, &error);
+    g_key_file_load_from_file (config, user_config_file, G_KEY_FILE_NONE, &error);
     if (error && !g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
-        g_warning ("Failed to load configuration from %s: %s", USER_CONFIG_FILE, error->message);
+        g_warning ("Failed to load configuration from %s: %s", user_config_file, error->message);
 
     gint minimum_uid = 500;
     if (g_key_file_has_key (config, "UserList", "minimum-uid", NULL))
-- 
2.42.1

